### PR TITLE
add deprecated for softmax_with_cross_entropy

### PR DESCRIPTION
**Pull request by chajchaj**
### PR types
Function optimization

### PR changes
APIs

### Describe
1、softmax_with_cross_entropy是历史原因保留的api，2.0下并不推荐用户使用，所以增加打印deprecated警告
2、完善nn/functional/loss.py中cross_entropy的英文文档
3、完善nn/layer/loss.py中CrossEntropyLoss的英文文档
4、修复softlabel为true，并指定weight参数时cross_entropy报错问题
5、python端加开关softmax_switch

### CODE REVIEW

File: python/paddle/fluid/layers/loss.py [hunk 1]
```diff
@@ -1162,7 +1162,8 @@ def softmax_with_cross_entropy(logits,
                                ignore_index=kIgnoreIndex,
                                numeric_stable_mode=True,
                                return_softmax=False,
-                               axis=-1):
+                               axis=-1,
+                               softmax_switch=True):
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/layers/loss.py]: softmax_with_cross_entropy是待废弃的api，不用升级，建议不修改这部分fluid下的python代码。
新功能可以在paddle.nn.functional.loss.py文件，通过调用底层core.ops.softmax_with_cross_entropy和append op的方式实现，避免对待废弃代码的依赖。

File: python/paddle/nn/functional/loss.py [hunk 2]
```diff
@@ -1119,94 +1137,203 @@ def cross_entropy(input,
                   reduction='mean',
                   soft_label=False,
                   axis=-1,
+                  softmax_switch=True,
```
Comment (XiaoguangHu01) [inline python/paddle/nn/functional/loss.py]: switch这个单词有转换的意思，在这里可能不太直观；
建议用use_softmax=True或者enable_softmax=True，类似use_cudnn，use_global_stat

File: python/paddle/nn/functional/loss.py [hunk 3]
```diff
@@ -1119,94 +1137,239 @@ def cross_entropy(input,
...
+                  use_softmax=True,
                   name=None):
     r"""
     This operator implements the cross entropy loss function with softmax. This function 
     combines the calculation of the softmax operation and the cross entropy loss function 
-    to provide a more numerically stable gradient.
-    Because this operator performs a softmax on logits internally, it expects
-    unscaled logits. This operator should not be used with the output of
-    softmax operator since that would produce incorrect results.
+    to provide a more numerically stable computing.
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py]: 加了use_softmax这个参数后，这一段描述也相应的更新一下吧。这里只描述了default的行为。

File: python/paddle/nn/functional/loss.py [hunk 4]
```diff
@@ -1119,94 +1137,239 @@ def cross_entropy(input,
...
-    one class with a probability of 1.0. Each sample in the batch will have a 
-    single label.
+    By default, this operator will calculate the mean of the result, and you can also affect 
+    the default behavior by using the reduction parameter. Please refer to the part of 
+    parameters for details.
 
-    The equation is as follows:
+    This operator can be used to calculate the softmax cross entropy loss with soft and hard labels.
+    Where, the hard labels mean the actual label value, 0, 1, 2, etc.  And the soft labels 
+    mean the probability of the actual label, 0.6, 0.8, 0.2, etc.
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py:1151]: 碰到过好几个用户说看不懂这里是在说什么意思。看看还能不能优化一下，至少体现出来hard label时的值的范围是[0, C-1]， soft label的时候给的李孜的和为1.0。

File: python/paddle/nn/functional/loss.py [hunk 5]
```diff
@@ -1119,94 +1137,239 @@ def cross_entropy(input,
...
 
-    1) Hard label (one-hot label, so every sample has exactly one class)
+    The calculation of this operator includes the following two steps.
 
-    .. math::
+    - **I.softmax cross entropy**
 
-        loss_j =  -\\text{logits}_{label_j} +
-        \\log\\left(\\sum_{i=0}^{K}\\exp(\\text{logits}_i)\\right), j = 1,..., K
+        1. Hard label (each sample can only be assigned into one category)
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py:1157]: use_softmax=False时的公式，是否要体现出来。

File: python/paddle/nn/functional/loss.py [hunk 6]
```diff
@@ -1119,94 +1137,239 @@ def cross_entropy(input,
...
-	    (N, D1, D2,..., Dk), k >= 1.
-        weight (Tensor, optional):a manual rescaling weight given to each class. 
+
+        - **input** (Tensor)
+
+            Input tensor, the data type is float32, float64. Shape is
+	    :math:`[N_1, N_2, ..., N_k, C]`, where C is number of classes ,  ``k >= 1`` . 
+            Note: it expects unscaled logits. This operator should not be used with the 
+            output of softmax operator, which will produce incorrect results.
+
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py:1257]: use_softmax = False时的输入，可以是softmax的结果。

File: python/paddle/nn/functional/loss.py [hunk 7]
```diff
@@ -1119,94 +1137,239 @@ def cross_entropy(input,
...
+            
             import paddle
             import numpy as np
+            np.random.seed(99999)
+            N=100
+            C=200
+            reduction='mean'
+            input_np = np.random.random([N, C]).astype(np.float64)  
+            label_np = np.random.randint(0, C, size=(N)).astype(np.int64)  
+            weight_np = np.random.random([C]).astype(np.float64)
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py]: 构造示例的时候可以直接用paddle.rand

File: python/paddle/nn/functional/loss.py [hunk 8]
```diff
@@ -1119,94 +1137,239 @@ def cross_entropy(input,
...
+
+
+            1.2. Soft labels (soft_label = True)
+
+             .. math::
+                \\loss_j=loss_j*\sum_{i}\left(weight[label_i]*logits_i\right)
+
+        2. reduction
+
+            2.1 if the ``reduction`` parameter is ``none``
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py:1215]: none建议改为None。

File: python/paddle/nn/layer/loss.py [hunk 9]
```diff
@@ -141,86 +141,234 @@ def forward(self, logit, label):
...
+            C = 3
+            shape = [N, C]
+            use_softmax = True
+            reduction='mean'
+            weight = None
+            logits = np.random.uniform(0.1, 1.0, shape).astype(dtype)
+            labels = np.random.uniform(0.1, 1.0, shape).astype(dtype)
+            labels /= np.sum(labels, axis=axis, keepdims=True)
+            paddle.set_device("cpu")
+            paddle.disable_static()
```
Comment (jzhang533) [inline python/paddle/nn/layer/loss.py]: 为什么需要调用`paddle.disable_static`?

**Review by jzhang533 — COMMENTED** [review]
文档中的问题还麻烦改进。

File: python/paddle/nn/layer/loss.py (hunk 9 above)
Comment (chajchaj) [reply python/paddle/nn/layer/loss.py]: 已修改

File: python/paddle/nn/functional/loss.py (hunk 8 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py:1215]: 已修改weight为None

File: python/paddle/nn/functional/loss.py (hunk 3 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py]: 已更新

File: python/paddle/nn/functional/loss.py (hunk 4 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py:1151]: 已更新

File: python/paddle/nn/functional/loss.py (hunk 7 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py]: 已修改

File: python/paddle/nn/functional/loss.py (hunk 6 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py:1257]: 已增加

File: python/paddle/nn/functional/loss.py (hunk 5 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py:1157]: 已增加

File: python/paddle/fluid/layers/loss.py (hunk 1 above)
Comment (chajchaj) [reply python/paddle/fluid/layers/loss.py]: 已修改

File: python/paddle/nn/functional/loss.py (hunk 2 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py]: 已修改

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

**Review by chalsliu — DISMISSED** [review]

File: python/paddle/nn/functional/loss.py [hunk 10]
```diff
@@ -1112,101 +1113,284 @@ def ctc_loss(log_probs,
...
-            label_data = np.random.randint(0, 100, size=(5)).astype(np.int64)
-            weight_data = np.random.random([100]).astype("float64")
-
-            input =  paddle.to_tensor(input_data)
-            label =  paddle.to_tensor(label_data)
-            weight = paddle.to_tensor(weight_data)
+        .. code-block:: python
+            
+            import paddle
+            import numpy as np
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py]: 这行不需要吧。

File: python/paddle/nn/functional/loss.py [hunk 11]
```diff
@@ -1112,101 +1113,284 @@ def ctc_loss(log_probs,
...
+            axis = -1
+            ignore_index = -100
+            N = 4
+            C = 3
+            shape = [N, C]
+            reduction='mean'
+            weight = None
+            logits = paddle.uniform(shape, dtype='float64', min=0.1, max=1.0, seed=99999)
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py]: 这里为啥没像example1一样，用paddle.seed设置种子呢？

File: python/paddle/nn/functional/loss.py [hunk 12]
```diff
@@ -1112,101 +1113,284 @@ def ctc_loss(log_probs,
...
+            labels = paddle.uniform(shape, dtype='float64', min=0.1, max=1.0, seed=99999)
+            labels /= paddle.sum(labels, axis=axis, keepdim=True)
+            paddle.set_device("cpu")
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py]: 这行是不是也不需要？

File: python/paddle/nn/functional/loss.py [hunk 13]
```diff
@@ -1112,101 +1113,284 @@ def ctc_loss(log_probs,
...
-        soft_label (bool): indicate whether label is soft. Default False, meaning that
-                the label is hard. If soft_label=True, the label is soft.
-        axis (int, optional): The index of dimension to perform softmax calculations. It 
-                              should be in range :math:`[-1, rank - 1]`, while :math:`rank`
-                              is the rank of input :attr:`logits`. Default: -1.
 
+        - **soft_label** (bool, optional)
+
+            Indicate whether label is soft. 
+            If soft_label=False, the label is hard.  If soft_label=True, the label is soft.
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py]: 这句话好奇怪呀。

File: python/paddle/nn/functional/loss.py [hunk 14]
```diff
@@ -1112,101 +1113,284 @@ def ctc_loss(log_probs,
...
+            Default is ``False``.
+
+        - **axis** (int, optional)
+
+            The index of dimension to perform softmax calculations. 
+            It should be in range :math:`[-1, rank - 1]`, where :math:`rank` is the rank of
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py]: rank 可改为number of dimensions 或者缩写ndim。（可参考API的规范）

**Review by raindrops2sea — DISMISSED** [review]

File: python/paddle/nn/functional/loss.py (hunk 14 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py]: 已更新

File: python/paddle/nn/functional/loss.py (hunk 12 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py]: 已更新

File: python/paddle/nn/functional/loss.py (hunk 10 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py]: 已更新

File: python/paddle/nn/functional/loss.py (hunk 11 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py]: 已更新

**Review by jzhang533 — DISMISSED** [review]
lgtm

File: python/paddle/nn/functional/loss.py (hunk 13 above)
Comment (chajchaj) [reply python/paddle/nn/functional/loss.py]: 已删

**Review by raindrops2sea — APPROVED** [review]

**Review by chalsliu — APPROVED** [review]
